File: backend/src/main/java/com/smartmeal/service/NotificationService.java
package com.smartmeal.service;

import com.google.firebase.messaging.FirebaseMessaging;
import com.google.firebase.messaging.Message;
import com.google.firebase.messaging.Notification;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

@Service
public class NotificationService {

    private static final Logger logger = LoggerFactory.getLogger(NotificationService.class);

    public void sendMealRecommendationNotification(Long userId, String userFcmToken, 
            String mealType, String itemName, String restaurantName) {
        if (userFcmToken == null || userFcmToken.isEmpty()) {
            logger.warn("FCM token not available for user {}", userId);
            return;
        }

        try {
            Message message = Message.builder()
                    .token(userFcmToken)
                    .notification(Notification.builder()
                            .setTitle("🍽️ Time for " + mealType + "!")
                            .setBody(itemName + " from " + restaurantName + " has been added to your cart")
                            .build())
                    .putData("type", "meal_recommendation")
                    .putData("userId", userId.toString())
                    .putData("mealType", mealType)
                    .putData("itemName", itemName)
                    .putData("restaurantName", restaurantName)
                    .build();

            String response = FirebaseMessaging.getInstance().send(message);
            logger.info("Successfully sent meal recommendation notification to user {}: {}", userId, response);
        } catch (Exception e) {
            logger.error("Failed to send notification to user {}: {}", userId, e.getMessage());
        }
    }

    public void sendOrderStatusNotification(Long userId, String userFcmToken, 
            String orderId, String status, String message) {
        if (userFcmToken == null || userFcmToken.isEmpty()) {
            logger.warn("FCM token not available for user {}", userId);
            return;
        }

        try {
            Message.Builder messageBuilder = Message.builder()
                    .token(userFcmToken)
                    .putData("type", "order_status")
                    .putData("orderId", orderId)
                    .putData("status", status);

            String title;
            String body;

            switch (status.toUpperCase()) {
                case "CONFIRMED":
                    title = "✅ Order Confirmed";
                    body = "Your order #" + orderId + " has been confirmed";
                    break;
                case "PREPARING":
                    title = "👨‍🍳 Preparing Your Order";
                    body = "Your order is being prepared";
                    break;
                case "OUT_FOR_DELIVERY":
                    title = "🚴 Out for Delivery";
                    body = "Your order is on its way";
                    break;
                case "DELIVERED":
                    title = "✅ Order Delivered";
                    body = "Your order has been delivered. Enjoy!";
                    break;
                case "CANCELLED":
                    title = "❌ Order Cancelled";
                    body = "Your order #" + orderId + " has been cancelled";
                    break;
                default:
                    title = "Order Update";
                    body = message != null ? message : "Your order status has been updated";
            }

            messageBuilder.notification(Notification.builder()
                    .setTitle(title)
                    .setBody(body)
                    .build());

            String response = FirebaseMessaging.getInstance().send(messageBuilder.build());
            logger.info("Successfully sent order status notification to user {}: {}", userId, response);
        } catch (Exception e) {
            logger.error("Failed to send order status notification to user {}: {}", userId, e.getMessage());
        }
    }

    public void sendSmaTriggerNotification(Long userId, String userFcmToken, 
            String mealType, String scheduledTime) {
        if (userFcmToken == null || userFcmToken.isEmpty()) {
            logger.warn("FCM token not available for user {}", userId);
            return;
        }

        try {
            Message message = Message.builder()
                    .token(userFcmToken)
                    .notification(Notification.builder()
                            .setTitle("⏰ " + mealType + " Time!")
                            .setBody("Your scheduled meal will be auto-added in 1 hour at " + scheduledTime)
                            .build())
                    .putData("type", "sma_reminder")
                    .putData("userId", userId.toString())
                    .putData("mealType", mealType)
                    .putData("scheduledTime", scheduledTime)
                    .build();

            String response = FirebaseMessaging.getInstance().send(message);
            logger.info("Successfully sent SMA reminder to user {}: {}", userId, response);
        } catch (Exception e) {
            logger.error("Failed to send SMA reminder to user {}: {}", userId, e.getMessage());
        }
    }

    public void sendGenericNotification(String token, String title, String body, java.util.Map<String, String> data) {
        if (token == null || token.isEmpty()) {
            return;
        }

        try {
            Message.Builder messageBuilder = Message.builder()
                    .token(token)
                    .notification(Notification.builder()
                            .setTitle(title)
                            .setBody(body)
                            .build());

            if (data != null) {
                for (java.util.Map.Entry<String, String> entry : data.entrySet()) {
                    messageBuilder.putData(entry.getKey(), entry.getValue());
                }
            }

            FirebaseMessaging.getInstance().send(messageBuilder.build());
        } catch (Exception e) {
            logger.error("Failed to send generic notification: {}", e.getMessage());
        }
    }
}

File: sma_app/lib/data/services/notification_service.dart
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:shared_preferences/shared_preferences.dart';

class NotificationService {
  static final NotificationService _instance = NotificationService._internal();
  factory NotificationService() => _instance;
  NotificationService._internal();

  final FlutterLocalNotificationsPlugin _notifications =
      FlutterLocalNotificationsPlugin();
  bool _isInitialized = false;

  Future<void> initialize() async {
    if (_isInitialized) return;

    const androidSettings =
        AndroidInitializationSettings('@mipmap/ic_launcher');
    const iosSettings = DarwinInitializationSettings(
      requestAlertPermission: true,
      requestBadgePermission: true,
      requestSoundPermission: true,
    );

    const initSettings = InitializationSettings(
      android: androidSettings,
      iOS: iosSettings,
    );

    await _notifications.initialize(
      initSettings,
      onDidReceiveNotificationResponse: _onNotificationTapped,
    );

    _isInitialized = true;
  }

  void _onNotificationTapped(NotificationResponse response) {
    final payload = response.payload;
    if (payload == null) return;

    debugPrint('Notification tapped with payload: $payload');
  }

  Future<bool> requestPermission() async {
    if (Platform.isAndroid) {
      final androidPlugin =
          _notifications.resolvePlatformSpecificImplementation<
              AndroidFlutterLocalNotificationsPlugin>();
      final granted = await androidPlugin?.requestNotificationsPermission();
      return granted ?? false;
    } else if (Platform.isIOS) {
      final iosPlugin = _notifications.resolvePlatformSpecificImplementation<
          IOSFlutterLocalNotificationsPlugin>();
      final granted = await iosPlugin?.requestPermissions(
        alert: true,
        badge: true,
        sound: true,
      );
      return granted ?? false;
    }
    return true;
  }

  Future<void> saveFcmToken(String token) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('fcm_token', token);
  }

  Future<String?> getFcmToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('fcm_token');
  }

  Future<void> showMealRecommendation({
    required String mealType,
    required String itemName,
    required String restaurantName,
  }) async {
    const androidDetails = AndroidNotificationDetails(
      'meal_recommendation',
      'Meal Recommendations',
      channelDescription: 'AI meal recommendations',
      importance: Importance.high,
      priority: Priority.high,
      icon: '@mipmap/ic_launcher',
    );

    const iosDetails = DarwinNotificationDetails(
      presentAlert: true,
      presentBadge: true,
      presentSound: true,
    );

    const details = NotificationDetails(
      android: androidDetails,
      iOS: iosDetails,
    );

    await _notifications.show(
      0,
      '🍽️ Time for $mealType!',
      '$itemName from $restaurantName has been added to your cart',
      details,
      payload: 'meal_recommendation',
    );
  }

  Future<void> showOrderStatus({
    required String orderId,
    required String status,
    String? message,
  }) async {
    String title;
    String body;

    switch (status.toUpperCase()) {
      case 'CONFIRMED':
        title = '✅ Order Confirmed';
        body = 'Your order #$orderId has been confirmed';
        break;
      case 'PREPARING':
        title = '👨‍🍳 Preparing Your Order';
        body = 'Your order is being prepared';
        break;
      case 'OUT_FOR_DELIVERY':
        title = '🚴 Out for Delivery';
        body = 'Your order is on its way';
        break;
      case 'DELIVERED':
        title = '✅ Order Delivered';
        body = 'Your order has been delivered. Enjoy!';
        break;
      case 'CANCELLED':
        title = '❌ Order Cancelled';
        body = 'Your order #$orderId has been cancelled';
        break;
      default:
        title = 'Order Update';
        body = message ?? 'Your order status has been updated';
    }

    const androidDetails = AndroidNotificationDetails(
      'order_status',
      'Order Status',
      channelDescription: 'Order status updates',
      importance: Importance.high,
      priority: Priority.high,
      icon: '@mipmap/ic_launcher',
    );

    const iosDetails = DarwinNotificationDetails(
      presentAlert: true,
      presentBadge: true,
      presentSound: true,
    );

    const details = NotificationDetails(
      android: androidDetails,
      iOS: iosDetails,
    );

    await _notifications.show(
      orderId.hashCode,
      title,
      body,
      details,
      payload: 'order_status:$orderId',
    );
  }

  Future<void> showSmaReminder({
    required String mealType,
    required String scheduledTime,
  }) async {
    const androidDetails = AndroidNotificationDetails(
      'sma_reminder',
      'SMA Reminders',
      channelDescription: 'Smart Meal Autopilot reminders',
      importance: Importance.high,
      priority: Priority.high,
      icon: '@mipmap/ic_launcher',
    );

    const iosDetails = DarwinNotificationDetails(
      presentAlert: true,
      presentBadge: true,
      presentSound: true,
    );

    const details = NotificationDetails(
      android: androidDetails,
      iOS: iosDetails,
    );

    await _notifications.show(
      1,
      '⏰ $mealType Time!',
      'Your scheduled meal will be auto-added in 1 hour at $scheduledTime',
      details,
      payload: 'sma_reminder',
    );
  }

  Future<void> cancelAllNotifications() async {
    await _notifications.cancelAll();
  }

  Future<void> cancelNotification(int id) async {
    await _notifications.cancel(id);
  }
}
